src/auth/mod.rs

```diff
@@ -0,0 +1,202 @@
+// file, you can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Types for parsing and authenticating HAWK headers.
+//! Matches the [Python logic](https://github.com/mozilla-services/tokenlib).
+//! We may want to extract this to its own repo/crate in due course.
+
+#[cfg(test)]
+mod test;
+
+use std::{
+    error::Error,
+    fmt::{self, Display, Formatter},
+};
+
+use actix_web::{error::ResponseError, http::header::ToStrError, FromRequest, HttpRequest};
+use base64::{self, DecodeError};
+use chrono::offset::Utc;
+use hawk::{Error as HawkError, Header as HawkHeader, Key, mac::{Mac as HawkMac, MacType}};
+use hkdf::{Hkdf, InvalidLength};
+use hmac::{
+    crypto_mac::{InvalidKeyLength, MacError},
+    Hmac, Mac,
+};
+use ring;
+use serde_json::{self, Error as JsonError};
+use sha2::Sha256;
+use time::Duration;
+
+use settings::Settings;
+
+#[derive(Debug, Deserialize)]
+pub struct HawkPayload {
+    expires: f64,
+    node: String,
+    salt: String,
+    uid: u64,
+}
+
+impl HawkPayload {
+    fn new(
+        header: &str,
+        method: &str,
+        host: &str,
+        port: u16,
+        path: &str,
+        settings: &Settings,
+        expiry: u64
+    ) -> AuthResult<HawkPayload> {
+        if &header[0..5] != "Hawk " {
+            return Err(AuthError);
+        }
+
+        // Extract id from hawk header
+        let header: HawkHeader = header[5..].parse()?;
+        let id = header.id.as_ref().ok_or(AuthError)?;
+
+        // Extract payload and signature from hawk id
+        let decoded_id = base64::decode_config(id, base64::URL_SAFE)?;
+        let payload_length = decoded_id.len() - 32;
+        let payload = &decoded_id[0..payload_length];
+        let signature = &decoded_id[payload_length..];
+
+        // Derive signing secret from master secret
+        let hkdf: Hkdf<Sha256> = Hkdf::extract(None, &settings.master_secret);
+        let mut signing_secret = [0u8; 32];
+        hkdf.expand(
+            b"services.mozilla.com/tokenlib/v1/signing",
+            &mut signing_secret,
+        )?;
+
+        // Verify signature
+        let mut expected_signature: Hmac<Sha256> = Hmac::new_varkey(&signing_secret)?;
+        expected_signature.input(&payload);
+        expected_signature.verify(signature)?;
+
+        // Parse payload JSON
+        let payload: HawkPayload = serde_json::from_slice(payload)?;
+        println!("{:?}", payload);
+
+        // Check payload is not expired
+        if (payload.expires.round() as u64) <= expiry {
+            return Err(AuthError);
+        }
+
+        // Derive token secret from master secret + salt
+        let hkdf: Hkdf<Sha256> = Hkdf::extract(
+            Some(payload.salt.as_bytes()),
+            &settings.master_secret,
+        );
+        let mut token_secret = [0u8; 32];
+        hkdf.expand(
+            format!("services.mozilla.com/tokenlib/v1/derive/{}", id).as_bytes(),
+            &mut token_secret,
+        )?;
```

Comment:
I missed it in my initial description, but re-reading the python `tokenlib` source code, I believe that `token_secret` here needs to be b64urlsafe encoded before using it to verify the mac below.  I'm having a bit of trouble building the project right now, but if I can get that working I'll put an encode step in here to check.

Reply:
I believe it is supposed to be base64encoded. 

https://github.com/mozilla-services/go-syncstorage/blob/master/token/token.go#L101

Reply:
🎉 Woop woop! It's validating successfully now, thanks for the help with this!